src/nvim/ex_cmds.h

```diff
@@ -19,13 +19,47 @@
 #define ECMD_LAST       (linenr_T)-1    /* use last position in all files */
 #define ECMD_ONE        (linenr_T)1     /* use first line */
 
+/// for cmdl_progress in live substitution
+typedef enum {
+  LS_NO_WD,                 /// state of cmdline when none words are typed
+                              /// ":%s" or ":%s/"
+  LS_ONE_WD,                /// state of cmd line when only pattern word began
+                              /// to be typed : ":%s/patt"
+  LS_TWO_SLASH_ONE_WD,      /// Second / has been typed but not the second
+                              /// word yet : "%s/pattern/"
+  LS_TWO_WD                /// state of cmd line when pattern has been completed
+                            /// and substitue is being typed : ":%s/pattern/sub"
+} LiveSub_state;
+
 /// Previous :substitute replacement string definition
 typedef struct {
   char *sub;            ///< Previous replacement string.
   Timestamp timestamp;  ///< Time when it was last set.
   list_T *additional_elements;  ///< Additional data left from ShaDa file.
 } SubReplacementString;
 
+/// Defs for live_sub functionality
+#define _noop(x)
+/// initializer for a list of match in a line
+KLIST_INIT(colnr_T, colnr_T, _noop)
+
+/// structure to backup and display matched lines in live_substitution
+typedef struct {
+  linenr_T lnum;
+  long nmatch;
+  char_u *line;
+  klist_t(colnr_T) *start_col;  // NOLINT(whitespace/cast)
+} matchedline_T;
+
+#define _dealloc_matchedline_T(x) \
+{ \
+  if (x->data.line) { xfree(x->data.line); } \
+  if (x->data.start_col) { kl_destroy(colnr_T, x->data.start_col); } \
+} \
```

Comment:
https://stackoverflow.com/questions/154136/why-use-apparently-meaningless-do-while-and-if-else-statements-in-c-c-macros